File: 프로그래머스/Lv3/정수 삼각형.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <cmath>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    vector<vector<int>> v;
    v.resize(triangle.size());
    int Index=0;
    v[0].push_back(triangle[0][0]);
    for(int i=1;i<triangle.size();i++){
        v[i].resize(triangle[i].size());
        for(int j=0;j<triangle[i].size();j++){
            
            if(j==0){
                v[i][j]=v[i-1][j]+triangle[i][j];
            }else if(j==triangle[i].size()-1){
                v[i][j]=v[i-1][j-1]+triangle[i][j];
            }else{
                v[i][j]=max(v[i-1][j-1],v[i-1][j])+triangle[i][j];
            }
            
        }
    }
    
    for(int i=0;i<v.size();i++){
        if(v[v.size()-1][i]>answer) answer=v[v.size()-1][i];
    }
    
    return answer;
}

File: 프로그래머스/Lv3/정수 삼각형.py
def solution(triangle):
    answer = 0
    arr=[]
    Size=len(triangle)
    
    for i in range(Size-1):
        temp=[]
        for j in range(i+1):
            temp.append(-1)
        arr.append(temp)
                   
    arr.append(triangle[Size-1])
    
    for i in range(Size-2,-1,-1):
        for j in range(0,i+1):
            arr[i][j]=max(arr[i+1][j],arr[i+1][j+1])+triangle[i][j]
    
    answer=arr[0][0]
    return answer
